### PR TITLE
Removing Sleep function from resource_group_membership.go

### PR DESCRIPTION
**Pull request by tmeckel**
## All Submissions:
-------------------------------------
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [NA] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open PRs for the same update/change?

## What is the current behavior?
-------------------------------------

Currently the implementation inside resource_group_membership.go uses `Sleep` to wait for DevOps changing the requested membership changes. 

Issue Number: N/A

## What is the new behavior?
-------------------------------------

The PR contains changes to the implementation inside `resource_group_membership.go` which utilizes the `WaitForState()` method, which provides a reliable way to wait for changes to become effective on an effected resource.

This PR contains additions to the documentation in `docs/contributing.md` which explains why using `Sleep` for waiting on state changes on (cloud) resources are considered bad practice.

## Does this introduce a breaking change?
-------------------------------------
- [NO]

## Any relevant logs, error output, etc?
-------------------------------------
(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

N/A

## Other information
-------------------------------------

N/A

### CODE REVIEW

File: azuredevops/resource_group_membership.go [hunk 1]
```diff
@@ -108,8 +137,35 @@ func resourceGroupMembershipUpdate(d *schema.ResourceData, m interface{}) error
...
+			state := "Waiting"
+			actualMemberships, err := getGroupMemberships(clients, group)
+			if err != nil {
+				return nil, "", fmt.Errorf("Error reading group memberships: %+v", err)
+			}
+			actualMembershipsSet, err := getGroupMembershipSet(actualMemberships)
+			if err != nil {
+				return nil, "", fmt.Errorf("Error converting membership list to set: %+v", err)
+			}
+			if actualMembershipsSet.Intersection(membersToAdd).Len() <= 0 &&
```
Comment (nmiodice) [inline azuredevops/resource_group_membership.go]: Can you help me understand the logic here? I'd have expected `actualMembershipsSet.Intersection(membersToAdd).Len()` to yield a length equal to `len(membersToAdd)`, as that would indicate that all members have been added. Am I reading this incorrectly?

Really cool approach here, by the way. I did not know about `StateChangeConf` and love the approach you took here.

Comment (tmeckel) [reply azuredevops/resource_group_membership.go]: @nmiodice good catch!!

While analyzing the issue I discovered that I overlooked a subtle issue with the hashcode functions that are in use when you use the `d.GetChange("members")` method and the hashcode function that I used to create the set of the current group members. 

I documented this in the new `waitForGroupMembershipChange` function which will reduce the code used in the `resourceGroupMembershipCreate` and `resourceGroupMembershipUpdate` functions.

**Comment by tmeckel** [comment]
@nmiodice the new pipeline step is failing because the various required environment variables are not set
image

**Comment by nmiodice** [comment]
@tmeckel @mikaelkrief 

Huh... I see now in the GH docs the following:
> Secrets are environment variables that are encrypted and only exposed to selected actions. Secrets are not passed to workflows that are triggered by a pull request from a fork of this repository. Learn more

I'll try to dive into this more when I'm back from holiday travels later this week. As of now, it is not a requirement for PRs so at least you get visibility into things like lint checks, etc...

**Comment by tmeckel** [comment]
@nmiodice Okay thanks for the clarification! Have great and safe travels!

**Pull request merged**